saltunnel/saltunnel-server.1
.TH saltunnel-client 1 "May 2020" "saltunnel" "saltunnel"

.SH NAME
.PP
\fBsaltunnel-client\fP, \fBsaltunnel-server\fP  - establish a cryptographically secure TCP tunnel

.SH SYNOPSIS
.PP
.B saltunnel-client [--version] [-t <timeout>] [-m <maxconns>] -k <keyfile> <fromip>:<fromport> <toip>:<toport>
.PP
.B saltunnel-server [--version] [-t <timeout>] [-m <maxconns>] -k <keyfile> <fromip>:<fromport> <toip>:<toport>

.SH DESCRIPTION
.PP
The \fBsaltunnel-client\fP and \fBsaltunnel-server\fP programs are used in conjunction to establish a cryptographically secure, difficult-to-censor, TCP tunnel using symmetric key cryptography.  This allows one to augment a normally-insecure TCP session with state-of-the-art security, with minimal hassle and minimal impact on performance. 

The following options are available:

    \fB--version\fP 
.RS 12
Display version information.
.RE

    \fB-t <timeout>\fP 
.RS 12
Set a connection timeout (in seconds) for all outgoing TCP connections. (Optional; defaults to system default.)
.RE

    \fB-m <maxconns>\fP 
.RS 12
Set a limit on the maximum number of concurrent TCP connections. (Optional; defaults to 100.)
.RE

    \fB-k <keyfile>\fP 
.RS 12
Specify the path to the shared key file. This file must be 32-bytes, and should be derived from a cryptographic random number generator.
.RE

    \fB<fromip>:<fromport>\fP 
.RS 12
The \fBip:port\fP pair identifying the source address to listen for incoming TCP connections. (For \fBsaltunnel-client\fP this is data arrives unencrypted; for \fBsaltunnel-server\fP it arrives encrypted.)
.RE

    \fB<toip>:<toport>\fP 
.RS 12
CRAP The \fBip:port\fP pair identifying the destination address to connect to for outgoing (encrypted) TCP connections. (For \fBsaltunnel-client\fP this is data is sent unencrypted; for \fBsaltunnel-server\fP it is sent encrypted.)
.RE


.SH EXAMPLES

Imagine that you have two computers on two different continents--ComputerA and ComputerB--and would like for some program on ComputerA to establish a TCP session with some port ComputerB, but the protocol which this program uses is too insecure to use comfortably over a network. You can use \fBsaltunnel\fP to easily solve this problem. Going further, imagine both of the countries your computers are in have border firewalls which detect and block common encryption protocols: \fBsaltunnel\fP would help here, too, as the \fBsaltunnel\fP protocol is difficult to detect.

First, you will need to generate a shared key and (somehow) copy it to both computers; \fBsaltunnel\fP requires 32-byte keys, which should be generated from a cryptographically secure source. 

So, on ComputerA, generate a key:

.RS 8
\fB<fromip>:<fromport>\fP 
mkdir ~/.saltunnel
cat /dev/urandom | head -c 32 > ~/.saltunnel/k
.RE

You'll then need to securely copy this to ComputerB.  (Doing this step securely is critical, but \fBsaltunnel\fP isn't designed to help with this step, so we will hand-wave this step. Fortunately, you'll only need to do this once.)

Next, you can set up the client-side of a tunnel. On ComputerA, you run:

.RS 8
saltunnel-client -k ~/.saltunnel/k 127.0.0.1:8080 computer-a:30000
.RE

(Where 'computer-a' is the hostname or IP address of ComputerA.)

After running this, any TCP connections made from ComputerA to IP address 127.0.0.1, port 7000, will be encrypted and forwarded to ComputerA's port 7000.

But, you'll still need to set up the server-side of the tunnel.  On ComputerB, you run:

.RS 8
saltunnel-server -k ~/.saltunnel/k 0.0.0.0:30000 127.0.0.1:8080
.RE

After running this, any TCP connections incoming to ComputerB on port 8080 will be decrypted and forwarded to IP address 127.0.0.1, port 8080.

Now that both ends of the tunnel are initialized, ComputerA can securely communicate with ComputerB. If ComputerA now makes a TCP connection to its own port 8080, the behavior should be as if it is connecting directly to ComputerB's local port 8080.

.PP
The \fIsecretkeyfile\fP should point to a file with the secret key.
.SH EXIT CODE
.PP
Normally, \fBsaltunnel-client\fP does not exit. (It will continue to listen for connections until terminated.)
.PP
If invalid arguments are given, \fBsaltunnel-client\fP exits with \fB2\fP.
.PP
If an error occurs, \fBsaltunnel-client\fP exits with \fB1\fP.
.SH CRYPTOGRAPHY
.PP
The protocol used by \fBsaltunnel\fP provides the following features:

.IP \[bu] 2
Confidentiality
.IP \[bu]
Integrity
.IP \[bu]
Post-Quantum Security
.IP \[bu]
Steganography
.IP \[bu]
Forward Secrecy
.IP \[bu]
Denial-of-Service Protection

.PP
The \fBsaltunnel\fP protocol uses salsa20+poly1305 as its primary symmetric cipher, which provides confidentiality and integrity. The protocol is also immune to quantum computing attacks, assuming the key is shared securely between parties (again due to the reliance on a symmetric cipher).  Additionally, it is steganographic, in that the entire protocol is indistinguishable from random: if someone attempts to eavesdrop on the conversation, they will be unable to detect any (for example) protocol header magic bytes or patterns in the data (apart from traffic analysis). Forward secrecy is also ensured: after a TCP connection terminates, it is impossible for even owners of the original shared key to decrypt it. Finally, \fBsaltunnel\fP offers considerable denial-of-service protection: an attacker attempting to overload a \fBsaltunnel-server\fP instance will find that his/her most cost-effective attacks are (probably) the well-known shortcomings of TCP itself.

.SH AUTHOR
Jay Sullivan <jay@identity.pub>
.SH "SEE ALSO"
saltunnel-server(1)


saltunnel/saltunnel-server.c
//
//  saltunnel-server.c
//  saltunnel
//

#include "src/saltunnel_tcp_server_forwarder.h"
#include "src/log.h"
#include "src/oops.h"
#include "src/rwn.h"
#include "src/math.h"
#include "src/threadpool.h"
#include "src/hypercounter.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sodium.h>

static void oops_usage() {
    fprintf(stderr, "saltunnel-server: usage: saltunnel-server [-t <timeout>] [-m <maxconns>] -k <keyfile> <fromip>:<fromport> <toip>:<toport>\n");
    exit(2);
}

static cache table = {0};

int main(int argc, char * argv[])
{
    // If version was requested, simply print it and exit
    if(argc>=2 && strcmp("--version",argv[1])==0) {
        printf("1.0.0\n");
        return 0;
    }
    
    // Parse arguments via getopt
    int opt;
    int verbosity = 0;
    const char* keyfile = 0;
    int keyfile_provided = 0;
    int maxconns = 100;

    while ((opt = getopt(argc, argv, "vt:m:k:")) != -1) {
        switch (opt) {
        case 'v':
            verbosity++;
            break;
        case 'k':
            keyfile = optarg;
            keyfile_provided = 1;
            break;
        case 'm':
            maxconns = atoi(optarg);
            break;
        case 't': {
            float connection_timeout_s ;
            if(sscanf(optarg, "%f", &connection_timeout_s)>=0 && connection_timeout_s>=0) {
                config_connection_timeout_ms = (int)(connection_timeout_s*1000);
            }
            break;
        }
        default:
            oops_usage();
        }
    }

    // Make sure we got exactly 2 positional args and 1 key
    int pos_arg_i = optind;
    int pos_arg_c = argc - optind;
    if(pos_arg_c != 2 || !keyfile)
        oops_usage();

    // Set verbosity level
    log_level = 2 - MAX(0,MIN(2,verbosity));
    
    // Set hard upper limit on number of connections allowed (it's really just a file descriptor limit)
    struct rlimit limit = { .rlim_cur=maxconns+4, .rlim_max=maxconns+4 };
    if(setrlimit(RLIMIT_NOFILE, &limit))
        oops_warn_sys("failed to limit number of file descriptors");

    // Parse [fromip]:[fromport]
    const char * first_arg_ptr = argv[pos_arg_i+0];
    char* from_colon = strchr(first_arg_ptr, ':');
    if(from_colon==0) oops_usage();
    *from_colon = 0;
    const char* from_host = first_arg_ptr;
    const char* from_port = from_colon+1;

    // Parse [toip]:[toport]
    const char * second_arg_ptr = argv[pos_arg_i+1];
    char* to_colon = strchr(second_arg_ptr, ':');
    if(to_colon==0) oops_usage();
    *to_colon = 0;
    const char* to_host = second_arg_ptr;
    const char* to_port = to_colon+1;

    // Read the key
    unsigned char key[32];
    if(mlock(key, sizeof(key))<0)
        oops_warn_sys("failed to mlock key");
    int key_fd = open(keyfile, O_RDONLY);
    if(key_fd<0)
        oops_error_sys("failed to open key");
    if(readn(key_fd, (char*)key,  sizeof(key))<0)
        oops_error_sys("failed to read key");
    if(close(key_fd)<0) oops_error_sys("failed to close fd");

    // Seed random bytes
    try(sodium_init())
    || oops_error("failed to initialize libsodium");
    
    // Initialize hypercounter
    try(hypercounter_init())
    || oops_error("failed to initialize hypercounter");
    
    // Initialize thread pool
    threadpool_init_all();

    // Run the server forwarder
    saltunnel_tcp_server_forwarder(&table, key, from_host, from_port, to_host, to_port);

    // The forwarder should never exit except in case of error
    return 1;
}
